// @material-ui/lab/esm/SpeedDialAction/SpeedDialAction.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
// @inheritedComponent Tooltip
import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { emphasize, withStyles } from '@material-ui/core/styles';
import Fab from '@material-ui/core/Fab';
import Tooltip from '@material-ui/core/Tooltip';
export var styles = function styles(theme) {
  return {
    /* Styles applied to the `Button` component. */
    button: {
      margin: 8,
      color: theme.palette.text.secondary,
      backgroundColor: theme.palette.common.white,
      '&:hover': {
        backgroundColor: emphasize(theme.palette.common.white, 0.15)
      },
      transition: "".concat(theme.transitions.create('transform', {
        duration: theme.transitions.duration.shorter
      }), ", opacity 0.8s"),
      opacity: 1
    },

    /* Styles applied to the `Button` component if `open={false}`. */
    buttonClosed: {
      opacity: 0,
      transform: 'scale(0)'
    }
  };
};
var SpeedDialAction = React.forwardRef(function SpeedDialAction(props, ref) {
  var ButtonProps = props.ButtonProps,
      classes = props.classes,
      className = props.className,
      _props$delay = props.delay,
      delay = _props$delay === void 0 ? 0 : _props$delay,
      icon = props.icon,
      id = props.id,
      onClick = props.onClick,
      onKeyDown = props.onKeyDown,
      _props$open = props.open,
      open = _props$open === void 0 ? false : _props$open,
      TooltipClasses = props.TooltipClasses,
      _props$tooltipOpen = props.tooltipOpen,
      tooltipOpenProp = _props$tooltipOpen === void 0 ? false : _props$tooltipOpen,
      _props$tooltipPlaceme = props.tooltipPlacement,
      tooltipPlacement = _props$tooltipPlaceme === void 0 ? 'left' : _props$tooltipPlaceme,
      tooltipTitle = props.tooltipTitle,
      other = _objectWithoutProperties(props, ["ButtonProps", "classes", "className", "delay", "icon", "id", "onClick", "onKeyDown", "open", "TooltipClasses", "tooltipOpen", "tooltipPlacement", "tooltipTitle"]);

  var _React$useState = React.useState(tooltipOpenProp),
      tooltipOpen = _React$useState[0],
      setTooltipOpen = _React$useState[1];

  var timeout = React.useRef();

  var _React$useState2 = React.useState(null),
      prevPropOpen = _React$useState2[0],
      setPreviousOpen = _React$useState2[1]; // getDerivedStateFromProps alternate


  if (!open && tooltipOpen) {
    setTooltipOpen(false);
    setPreviousOpen(open);
  }

  React.useEffect(function () {
    if (!tooltipOpenProp || prevPropOpen === open) {
      return undefined;
    }

    if (!tooltipOpen) {
      timeout.current = setTimeout(function () {
        return setTooltipOpen(true);
      }, delay + 100);
      return function () {
        clearTimeout(timeout.current);
      };
    }

    return undefined;
  });

  var handleTooltipClose = function handleTooltipClose() {
    if (tooltipOpenProp) return;
    setTooltipOpen(false);
  };

  var handleTooltipOpen = function handleTooltipOpen() {
    if (tooltipOpenProp) return;
    setTooltipOpen(true);
  };

  var clickProp = {
    onClick: onClick
  };

  if (typeof document !== 'undefined' && 'ontouchstart' in document.documentElement) {
    var startTime;
    clickProp = {
      onTouchStart: function onTouchStart() {
        startTime = new Date();
      },
      onTouchEnd: function onTouchEnd(event) {
        // only perform action if the touch is a tap, i.e. not long press
        if (new Date() - startTime < 500) {
          onClick(event);
        }
      }
    };
  }

  return React.createElement(Tooltip, _extends({
    id: id,
    ref: ref,
    title: tooltipTitle,
    placement: tooltipPlacement,
    onClose: handleTooltipClose,
    onOpen: handleTooltipOpen,
    open: open && tooltipOpen,
    classes: TooltipClasses
  }, other), React.createElement(Fab, _extends({
    size: "small",
    className: clsx(className, classes.button, !open && classes.buttonClosed),
    style: {
      transitionDelay: "".concat(delay, "ms")
    },
    tabIndex: -1,
    role: "menuitem",
    onKeyDown: onKeyDown
  }, ButtonProps, clickProp), icon));
});
process.env.NODE_ENV !== "production" ? SpeedDialAction.propTypes = {
  /**
   * Props applied to the [`Button`](/api/button/) component.
   */
  ButtonProps: PropTypes.object,

  /**
   * Override or extend the styles applied to the component.
   * See [CSS API](#css) below for more details.
   */
  classes: PropTypes.object.isRequired,

  /**
   * @ignore
   */
  className: PropTypes.string,

  /**
   * Adds a transition delay, to allow a series of SpeedDialActions to be animated.
   */
  delay: PropTypes.number,

  /**
   * The Icon to display in the SpeedDial Floating Action Button.
   */
  icon: PropTypes.node.isRequired,

  /**
   * @ignore
   */
  id: PropTypes.string,

  /**
   * @ignore
   */
  onClick: PropTypes.func,

  /**
   * @ignore
   */
  onKeyDown: PropTypes.func,

  /**
   * @ignore
   */
  open: PropTypes.bool,

  /**
   * Classes applied to the [`Tooltip`](/api/tooltip/) element.
   */
  TooltipClasses: PropTypes.object,

  /**
   * Make the tooltip always visible when the SpeedDial is open.
   */
  tooltipOpen: PropTypes.bool,

  /**
   * Placement of the tooltip.
   */
  tooltipPlacement: PropTypes.oneOf(['bottom-end', 'bottom-start', 'bottom', 'left-end', 'left-start', 'left', 'right-end', 'right-start', 'right', 'top-end', 'top-start', 'top']),

  /**
   * Label to display in the tooltip.
   */
  tooltipTitle: PropTypes.node.isRequired
} : void 0;
export default withStyles(styles, {
  name: 'MuiSpeedDialAction'
})(SpeedDialAction);
